### PR TITLE
feat(ec2): add options for Cfn init proxies

### LINKED ISSUE #19479: (ec2): allow --http-proxy and --https-proxy in Cloudformation helper scripts
### Description

[Cloudformation helper scripts](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-helper-scripts-reference.html) such as [`cfn-init`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-init.html) include options that allow users to set proxies.

```zsh
/opt/aws/bin/cfn-init --https-proxy https://my-proxy.com
```

These options aren't available in [ApplyCloudFormationInitOptions](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_autoscaling.ApplyCloudFormationInitOptions.html)

### Use Case

When operating in locked down networks where proxies are required to reach Cloudformation endpoints, the `--http-proxy` and/or `--https-proxy` options are necessary

### Proposed Solution

In [ApplyCloudFormationInitOptions](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_autoscaling.ApplyCloudFormationInitOptions.html), allow the option for `httpProxy` and `httpsProxy`

### Other information

I have a PR ready

### Acknowledge

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by maafk**
Closes #19479 

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](../CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](../CONTRIBUTING.md/#adding-new-unconventional-dependencies)

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by maafk** [comment]
@corymhall I must not be understanding the `prlint` failure I'm getting

> Error: Features must contain a change to an integration test file

I've tried adding and altering integration tests and continue to get this error. Am I missing something simple?

Thanks!

**Comment by peterwoodworth** [comment]
@maafk I've put in a fix for this https://github.com/aws/aws-cdk/pull/19521, hang tight!

**Comment by maafk** [comment]
Integration test was updated, and erroring due to a different orders in an IAM policy.

```
@aws-cdk/aws-ec2: Verifying integ.instance-init.js against integ.instance-init.expected.json ... CHANGED.
@aws-cdk/aws-ec2: Resources
@aws-cdk/aws-ec2: [~] AWS::IAM::Policy Instance2InstanceRoleDefaultPolicy610B37CD 
@aws-cdk/aws-ec2:  â””â”€ [~] PolicyDocument
@aws-cdk/aws-ec2:      â””â”€ [~] .Statement:
@aws-cdk/aws-ec2:          â””â”€ @@ -1,8 +1,8 @@
@aws-cdk/aws-ec2:             [ ] [
@aws-cdk/aws-ec2:             [ ]   {
@aws-cdk/aws-ec2:             [ ]     "Action": [
@aws-cdk/aws-ec2:             [+]       "s3:GetBucket*",
@aws-cdk/aws-ec2:             [ ]       "s3:GetObject*",
@aws-cdk/aws-ec2:             [-]       "s3:GetBucket*",
@aws-cdk/aws-ec2:             [ ]       "s3:List*"
@aws-cdk/aws-ec2:             [ ]     ],
@aws-cdk/aws-ec2:             [ ]     "Effect": "Allow",
@aws-cdk/aws-ec2:             @@ -18,7 +18,8 @@
@aws-cdk/aws-ec2:             [ ]       ":s3:::",
@aws-cdk/aws-ec2:             [ ]       {
@aws-cdk/aws-ec2:             [ ]         "Ref": "Asset1S3Bucket"
@aws-cdk/aws-ec2:             [-]       }
@aws-cdk/aws-ec2:             [+]       },
@aws-cdk/aws-ec2:             [+]       "/*"
@aws-cdk/aws-ec2:             [ ]     ]
@aws-cdk/aws-ec2:             [ ]   ]
@aws-cdk/aws-ec2:             [ ] },
@aws-cdk/aws-ec2:             @@ -33,8 +34,7 @@
@aws-cdk/aws-ec2:             [ ]       ":s3:::",
@aws-cdk/aws-ec2:             [ ]       {
@aws-cdk/aws-ec2:             [ ]         "Ref": "Asset1S3Bucket"
@aws-cdk/aws-ec2:             [-]       },
@aws-cdk/aws-ec2:             [-]       "/*"
@aws-cdk/aws-ec2:             [+]       }
@aws-cdk/aws-ec2:             [ ]     ]
@aws-cdk/aws-ec2:             [ ]   ]
@aws-cdk/aws-ec2:             [ ] }
```

These are exactly the same policies, but synthesized with actions/resources in a different order

**Comment by maafk** [comment]
@corymhall Anything I can clarify on this PR?

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 1]
```diff
@@ -1837,11 +1839,28 @@ export interface ApplyCloudFormationInitOptions {
...
+  readonly httpProxy?: string;
+
+  /**
+   * Include --https-proxy argument when running cfn-init and cfn-signal commands
+   *
+   * An HTTPS proxy. Use the following format: https://user:password@host:port
+   *
+   * @default
+   */
+  readonly httpsProxy?: string;
```
Comment (LukvonStrom) [inline packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1865]: as these are secrets, I would like to propose an alternative way:
An interface like this: 
```
interface IProxyConfig {
protocol: string,
user: string,
password: ISecret,
host: string,
port: number
}

Comment (maafk) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1865]: Finally getting back to this. Currently figuring out how to deal with circular dependencies between packages.

`aws-secretsmanager` has a dependency on the `aws-ec2` package, causing issues when doing a `yarn build` within the `aws-ec2` package after adding secrets manager.

Trying to figure out how to handle this if possible

Comment (maafk) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts:1865]: Articulated this in #19479

**Pull request closed without merging**